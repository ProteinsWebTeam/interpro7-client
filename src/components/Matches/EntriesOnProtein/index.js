/* eslint no-magic-numbers: 0 */
import React, {PropTypes as T} from 'react';
import ColorHash from 'color-hash/lib/color-hash';

import style from '../style.css';

const colorHash = new ColorHash();

const EntriesOnProtein = (
  {
    matches, options: {baseSize = 10, offset = 30, niceRatio = 6} = {},
  }
) => {
  const protein = matches[0].protein;
  protein.length = protein.length | matches[0].coordinates[0].protein[1];
  // TODO: make sure we have protein length in the coordinates object
  return (
    <div className={style.svgContainer}>
      <svg
        className={style.svg}
        preserveAspectRatio="xMinYMid meet"
        viewBox={`0 0 ${protein.length + offset} 60`}
      >
        <g transform={`translate(0 ${offset - baseSize / 2})`}>
            <title>{protein.accession}</title>
            <rect
              x="0" y="0" rx={baseSize / niceRatio}
              width={protein.length} height={baseSize}
              className={style.primary}
            />
          <text
            x="0.1em"
            y="0.8em"
            transform={`translate(${protein.length} 0)`}
          >
            <tspan>
              {protein.length}
            </tspan>
          </text>
        </g>
        <g>
          {
            matches.map(({coordinates: coords, entry}) => (
              <g
                key={entry.accession}
                transform={
                  `translate(${coords[0].protein[0]} ${offset - baseSize})`
                }
              >
                  <title>{entry.accession}</title>
                  <rect
                    x="0" y="0" rx={baseSize * 2 / niceRatio}
                    width={coords[0].protein[1] - coords[0].protein[0]}
                    fill={colorHash.hex(entry.accession)}
                    height={baseSize * 2}
                    className={style.secondary}
                  />
                <text y="-0.2em">
                  <tspan textAnchor="middle">
                    {coords[0].protein[0]}
                  </tspan>
                </text>
                <text
                  y="-0.2em"
                  transform={
                    `translate(${
                      coords[0].protein[1] - coords[0].protein[0]
                    } 0)`
                  }
                >
                  <tspan textAnchor="middle">
                    {coords[0].protein[1]}
                  </tspan>
                </text>
              </g>
            ))
          }
        </g>
      </svg>
    </div>
  );
};
EntriesOnProtein.propTypes = {
  matches: T.arrayOf(T.shape({
    coordinates: T.arrayOf(T.object).isRequired,
    protein: T.object.isRequired,
    entry: T.object.isRequired,
  })).isRequired,
  options: T.object,
};

export default EntriesOnProtein;
