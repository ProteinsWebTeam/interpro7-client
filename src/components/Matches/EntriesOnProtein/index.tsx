import React, { useEffect, useState } from 'react';

import { EntryColorMode, getTrackColor } from 'utils/entry-color';

import TooltipForTrack from 'components/SimpleCommonComponents/Tooltip/ForTrack';
import NightingaleInterProTrack from 'components/Nightingale/InterProTrack';
import NightingaleSequence from 'components/Nightingale/Sequence';
import { GenericMatch, Feature } from '../MatchesByPrimary';

import cssBinder from 'styles/cssBinder';

import protvista from 'components/ProteinViewer/style.css';
import EntryPopup from 'components/SimpleCommonComponents/Tooltip/EntryPopup';

const css = cssBinder(protvista);

type Props = { match: GenericMatch; matches: Array<AnyMatch> };

const EntriesOnProtein = ({ matches, match }: Props) => {
  const [data, setData] = useState<Array<Feature> | null>(null);
  const [locationHovered, setLocationHovered] = useState<
    Array<ProtVistaLocation>
  >([]);
  const { entry, protein } = match || {};

  useEffect(() => {
    if (matches === undefined || !matches.length || !entry || !protein) return;

    setData(
      matches.map((loc) => ({
        accession: entry.accession,
        name: entry.name,
        source_database: entry.source_database,
        locations: loc.entry_protein_locations || [],
        color: getTrackColor(entry, EntryColorMode.ACCESSION),
        entry_type: entry.entry_type,
        type: 'entry',
      }))
    );
  }, [entry, protein]);

  if (matches === undefined || !matches.length || !entry || !protein)
    return null;

  return (
    <div className={css('track-in-table')}>
      <div className={css('track-container')}>
        <div className={css('aligned-to-track-component')}>
          <NightingaleSequence
            sequence={protein.sequence || '\u00A0'.repeat(protein.length)}
            length={protein.length}
            display-start={1}
            display-end={protein.length}
            height={30}
            use-ctrl-to-zoom
          />
        </div>
      </div>
      <div className={css('track-component')}>
        <TooltipForTrack
          message={
            <EntryPopup
              locations={locationHovered}
              accession={entry.accession}
              dbName={entry.source_database}
              name={
                (entry?.name as NameObject)?.short ||
                (entry.name as string) ||
                ''
              }
            />
          }
          onMouseOverFeature={(locations: Array<ProtVistaLocation>) => {
            setLocationHovered(locations);
          }}
        >
          <NightingaleInterProTrack
            length={protein.length}
            display-start={1}
            display-end={protein.length}
            margin-color="#fafafa"
            id={`track_${entry.accession}`}
            data={data || []}
            shape="roundRectangle"
            expanded
            use-ctrl-to-zoom
          />
        </TooltipForTrack>
      </div>
    </div>
  );
};

export default EntriesOnProtein;
