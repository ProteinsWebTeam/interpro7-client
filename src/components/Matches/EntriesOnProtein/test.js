import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import EntriesOnProtein from '.';

const renderer = new ShallowRenderer();

describe('<EntriesOnProtein />', () => {
  test('should render', () => {
    renderer.render(
      <EntriesOnProtein
        matches={[
          {
            entry_protein_locations: [
              {
                fragments: [
                  {
                    end: 380,
                    start: 300,
                  },
                ],
              },
            ],
          },
        ]}
        match={{
          entry: {
            accession: 'IPR000001',
            name: {
              name: 'Kringle',
              short: 'Kringle',
            },
            source_database: 'interpro',
            type: 'domain',
          },
          protein: {
            accession: 'A0A016WA74',
            name: 'Uncharacterized protein',
            length: 882,
            protein_length: 882,
            source_database: 'unreviewed',
          },
        }}
      />,
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
