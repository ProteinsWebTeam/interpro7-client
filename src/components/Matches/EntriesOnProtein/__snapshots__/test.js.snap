// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EntriesOnProtein /> should render 1`] = `
<div
  className="track-in-table"
>
  <div
    className="track-container"
  >
    <div
      className="aligned-to-track-component"
    >
      <nightingale-sequence
        display-end={882}
        display-start="1"
        height="30"
        length={882}
        use-ctrl-to-zoom={true}
      />
    </div>
  </div>
  <div
    className="track-component"
  >
    <Tooltip
      title="<b>IPR000001</b><br/>Kringle<br/><p>300-380</p>"
    >
      <nightingale-interpro-track
        display-end={882}
        display-start="1"
        expanded={true}
        id="track_IPR000001"
        length={882}
        shape="roundRectangle"
        use-ctrl-to-zoom={true}
      />
    </Tooltip>
  </div>
</div>
`;
